hail/src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -986,343 +935,224 @@ object EmitPackDecoder {
   }
 }
 
-//
-// Generate the Type-specific C++ code for a PackDecoder
-//
-object NativeDecode {
-
-  def appendCode(sb: StringBuilder, rowType: PType, wantType: PType): Unit = {
-    val verbose = false
-    var seen = new ArrayBuffer[Int]()
-    val localDefs = new StringBuilder()
-    val entryCode = new StringBuilder()
-    val mainCode = new StringBuilder()
-
-    def stateVarType(name: String): String = {
-      name match {
-        case "len" => "ssize_t"
-        case "idx" => "ssize_t"
-        case "miss" => "std::vector<char>"
-        case _ => "char*"
-      }
-    }
+object NativeDecoder {
+  def skipBinary(input_buf_ptr: cxx.Expression): cxx.Code = {
+    val len = cxx.Variable("skip_len", "int", s"$input_buf_ptr->read_int()")
+    s"""${len.define}
+       |$input_buf_ptr->skip_bytes($len);""".stripMargin
+  }
 
-    def stateVar(name: String, depth: Int): String = {
-      val bit = name match {
-        case "len"  => 0x01
-        case "idx"  => 0x02
-        case "addr" => 0x04
-        case "ptr"  => 0x08
-        case "data" => 0x10
-        case "miss" => 0x20
-      }
-      if (seen.length <= depth) seen = seen.padTo(depth+1, 0)
-      val result = s"${name}${depth}"
-      if ((seen(depth) & bit) == 0) {
-        seen(depth) = (seen(depth) | bit)
-        val typ = stateVarType(name)
-        val initStr =
-          if (typ.equals("std::vector<char>")) ""
-          else if (!typ.equals("char*")) " = 0"
-          else " = nullptr"
-        localDefs.append(s"${typ} ${result}${initStr};\n")
-      }
-      result
+  def skipArray(t: PArray, input_buf_ptr: cxx.Expression): cxx.Code = {
+    val len = cxx.Variable("len", "int", s"$input_buf_ptr->read_int()")
+    val i = cxx.Variable("i", "int", "0")
+    if (t.elementType.required) {
+      s"""${len.define}
+         |for (${i.define} $i < $len; $i++) {
+         |  ${skip(t.elementType, input_buf_ptr)}
+         |}""".stripMargin
+    } else {
+      val missingBytes = cxx.ArrayVariable(s"missing", "char", s"n_missing_bytes($len)")
+      s"""${len.define}
+         |${missingBytes.define}
```

Comment:
Normally we have spaces inside the curly braces, so this will get a massive reformatting if it gets auto-indented.  Can we just do that now?